pkg/permissionclaim/permissionclaim_labeler.go

```diff
@@ -0,0 +1,94 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package permissionclaim
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/kcp-dev/logicalcluster/v2"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/clusters"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1/permissionclaims"
+	apisinformers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/indexers"
+)
+
+// Labeler calculates labels to apply to all instances of a cluster-group-resource based on permission claims.
+type Labeler struct {
+	listAPIBindingsAcceptingClaimedGroupResource func(clusterName logicalcluster.Name, groupResource schema.GroupResource) ([]*apisv1alpha1.APIBinding, error)
+	getAPIExport                                 func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIExport, error)
+}
+
+// NewLabeler returns a new Labeler.
+func NewLabeler(
+	apiBindingInformer apisinformers.APIBindingInformer,
+	apiExportInformer apisinformers.APIExportInformer,
+) *Labeler {
+	return &Labeler{
+		listAPIBindingsAcceptingClaimedGroupResource: func(clusterName logicalcluster.Name, groupResource schema.GroupResource) ([]*apisv1alpha1.APIBinding, error) {
+			indexKey := fmt.Sprintf("%s|%s", clusterName, groupResource)
+			return indexers.ByIndex[*apisv1alpha1.APIBinding](apiBindingInformer.Informer().GetIndexer(), indexers.APIBindingByClusterAndAcceptedClaimedGroupResources, indexKey)
+		},
+
+		getAPIExport: func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIExport, error) {
+			key := clusters.ToClusterAwareKey(clusterName, name)
+			return apiExportInformer.Lister().Get(key)
+		},
+	}
+}
+
+// LabelsFor returns all the applicable labels for the cluster-group-resource relating to permission claims. This is
+// the intersection of (1) all APIBindings in the cluster that have accepted claims for the group-resource with (2)
+// associated APIExports that are claiming group-resource.
+func (l *Labeler) LabelsFor(_ context.Context, cluster logicalcluster.Name, groupResource schema.GroupResource) (map[string]string, error) {
+	labels := map[string]string{}
+
+	bindings, err := l.listAPIBindingsAcceptingClaimedGroupResource(cluster, groupResource)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, binding := range bindings {
+		boundAPIExportWorkspace := binding.Status.BoundAPIExport.Workspace
+		export, err := l.getAPIExport(logicalcluster.New(boundAPIExportWorkspace.Path), boundAPIExportWorkspace.ExportName)
+		if err != nil {
+			// TODO(ncdc): should we (1) ignore errors, (2) return immediately, or (3) aggregate errors and return at
```

Comment:
as long as we use informers like we do with single shard this is clear indication that the export is gone. Then it would make sense to clean up the labels. In multiple shard case, we might just have lost connection or replication is slow or whatever. We would have to  make sure with a client call likely what is going on. Whatever we do here for now is fine. We will come back.

cc @p0lyn0mial

Reply:
we will also move part of this into a second controller. Then this lookup will go away, out of the critical path.

Reply:
Updated error handling logic, ptal